m27q.py

```diff
@@ -75,44 +110,53 @@ def set_osd(self, data: bytearray):
             message=bytearray([0x6E, 0x51, 0x81 + len(data), 0x03] + data),
         )
 
-    def set_brightness(self, brightness: int):
-        self.set_osd(
-            [
-                0x10,
-                0x00,
-                max(self._min_brightness, min(self._max_brightness, brightness)),
-            ]
-        )
+    def set_property(self, property: Property, value: int):
+        self.set_osd([property.message_a, property.message_b, property.clamp(value)])
 
-    def get_brightness(self):
-        return self.get_osd([0x10])
+    def get_property(self, property: Property):
+        return self.get_osd([property.message_a, property.message_b])
 
-    def transition_brightness(self, to_brightness: int, step: int = 3):
-        current_brightness = self.get_brightness()
-        diff = abs(to_brightness - current_brightness)
-        if current_brightness <= to_brightness:
+    def transition_property(self, property: BasicProperty, target: int, step: int = 3):
+        current = self.get_property(property)
+        diff = abs(target - current)
+        if current <= target:
             step = 1 * step  # increase
         else:
             step = -1 * step  # decrease
         while diff >= abs(step):
-            current_brightness += step
-            self.set_brightness(current_brightness)
+            current += step
+            self.set_property(property, current)
             diff -= abs(step)
         # Set one last time
-        if current_brightness != to_brightness:
-            self.set_brightness(to_brightness)
+        if current != target:
+            self.set_property(property, target)
+
+    def set_brightness(self, brightness: int):
```

Comment:
With this new abstraction perhaps we could leverage `__getattr__` to avoid implementing all these methods, e.g.
~~~ python
def __getattr__(self, name):
    action, _, property_name = name.partition("_")
    property_obj = getattr(MonitorControl, property_name.upper())
    if action == "get":
        return lambda: self.get_property(property_obj)
    elif action == "set":
        return lambda v: self.set_property(property_obj, v)

    raise AttributeError()
~~~

Reply:
I'll look at doing this when I get a chance.